runtimes/common/src/dip.rs

```diff
@@ -0,0 +1,698 @@
+	where
+		BlockNumber: MaxEncodedLen,
+	{
+		fn from(value: Vec<ProofEntry<BlockNumber>>) -> Self {
+			Self(value)
+		}
+	}
+
+	pub struct DidMerkleProofVerifier<KeyId, BlockNumber, Hasher>(PhantomData<(KeyId, BlockNumber, Hasher)>);
+
+	impl<KeyId, BlockNumber, Hasher> IdentityProofVerifier for DidMerkleProofVerifier<KeyId, BlockNumber, Hasher>
+	where
+		KeyId: MaxEncodedLen + Clone + Ord,
+		BlockNumber: MaxEncodedLen + Clone + Ord,
+		Hasher: sp_core::Hasher,
+	{
+		type BlindedValue = Vec<sender::BlindedValue>;
+		// TODO: Proper error handling
+		type Error = ();
+		type ProofDigest = <Hasher as sp_core::Hasher>::Out;
+		type ProofLeaf = ProofLeaf<KeyId, BlockNumber>;
+		type VerificationResult = VerificationResult<BlockNumber>;
+
+		fn verify_proof_against_digest(
+			proof: VersionedIdentityProof<Self::BlindedValue, Self::ProofLeaf>,
+			digest: Self::ProofDigest,
+		) -> Result<Self::VerificationResult, Self::Error> {
+			use dip_support::v1;
+			use sp_trie::verify_trie_proof;
+
+			let proof: v1::Proof<_, _> = proof.try_into()?;
+			// TODO: more efficient by removing cloning and/or collecting.
+			// Did not find another way of mapping a Vec<(Vec<u8>, Vec<u8>)> to a
+			// Vec<(Vec<u8>, Option<Vec<u8>>)>.
+			let proof_leaves = proof
+				.revealed
+				.iter()
+				.map(|leaf| (leaf.encoded_key(), Some(leaf.encoded_value())))
+				.collect::<Vec<(Vec<u8>, Option<Vec<u8>>)>>();
+			verify_trie_proof::<LayoutV1<Hasher>, _, _, _>(&digest, &proof.blinded, &proof_leaves).map_err(|_| ())?;
+
+			// At this point, we know the proof is valid. We just need to map the revealed
+			// leaves to something the consumer can easily operate on.
+
+			// Create a map of the revealed public keys
+			//TODO: Avoid cloning, and use a map of references for the lookup
+			let public_keys: BTreeMap<KeyId, DidPublicKeyDetails<BlockNumber>> = proof
+				.revealed
+				.clone()
+				.into_iter()
+				.filter_map(|leaf| {
+					if let ProofLeaf::KeyDetails(KeyDetailsKey(key_id), KeyDetailsValue(key_details)) = leaf {
+						Some((key_id, key_details))
+					} else {
+						None
+					}
+				})
+				.collect();
+			// Create a list of the revealed keys by consuming the provided key reference
+			// leaves, and looking up the full details from the just-built `public_keys`
+			// map.
+			let keys: Vec<ProofEntry<BlockNumber>> = proof
+				.revealed
+				.into_iter()
+				.filter_map(|leaf| {
+					if let ProofLeaf::KeyReference(KeyReferenceKey(key_id, key_relationship), KeyReferenceValue) = leaf
+					{
+						// TODO: Better error handling.
+						let key_details = public_keys
+							.get(&key_id)
+							.expect("Key ID should be present in the map of revealed public keys.");
+						Some(ProofEntry {
+							key: key_details.clone(),
+							relationship: key_relationship,
+						})
+					} else {
+						None
+					}
+				})
+				.collect();
+			Ok(keys.into())
+		}
+	}
+}
+
+#[cfg(test)]
```

Comment:
I would suggest to move these to a separate file.

Reply:
https://github.com/KILTprotocol/kilt-node/pull/492/commits/a2462e010fe73f171db1fe0c843f42ef3478a0e2